cli/azd/pkg/input/previewer.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package input
+
+import (
+	"strings"
+
+	tm "github.com/buger/goterm"
+)
+
+type previewer struct {
+	title        string
+	displayTitle string
+	footerLine   string
+	lines        int
+	prefix       string
+	output       []string
+}
+
+func NewPreviewer(lines int, prefix, title string) *previewer {
+	return &previewer{
+		lines:  lines,
+		prefix: prefix,
+		title:  title,
+	}
+}
+
+func (p *previewer) Start() {
+	p.output = make([]string, p.lines)
+	// title is created on Start() because it depends on terminal width
+	// if terminal is resized between stop and start, the previewer will
+	// react to it and update the size.
+	p.buildTitleMargins()
+	p.printOutput()
+}
+
+func clearLine(text string) {
+	sizeLog := len(text)
+	if sizeLog == 0 {
+		return
+	}
+	maxWidth := tm.Width()
+	if sizeLog > maxWidth {
+		sizeLog = maxWidth
+	}
+	eraseWith := strings.Repeat(" ", sizeLog)
+	tm.Printf(eraseWith)
+	tm.MoveCursorBackward(sizeLog)
+}
+
+func (p *previewer) Stop() {
+	p.clear()
+	p.output = nil
+}
+
+func (p *previewer) clear() {
+	if p.output == nil {
+		return
+	}
+
+	tm.MoveCursorBackward(len(p.footerLine))
+	clearLine(p.footerLine)
+	// skip empty line between footer and log
+	tm.MoveCursorUp(1)
+	size := len(p.output) - 1
+	for index := range p.output {
+		tm.MoveCursorUp(1)
+		clearLine(p.output[size-index])
+	}
+	// moving up two times, one is the empty line between title and logs
+	tm.MoveCursorUp(1)
+	tm.MoveCursorUp(1)
+	clearLine(p.displayTitle)
+
+	tm.Flush()
+}
+
+func (p *previewer) Write(logBytes []byte) (int, error) {
+	fullText := strings.Split(string(logBytes), "\n")
+	maxWidth := tm.Width()
+	if maxWidth <= 0 {
+		// tm.Width <= 0 means a CI terminal, where logs can be just written
+		return tm.Println(fullText)
+	}
+
+	for _, log := range fullText {
+		if log == "" {
+			continue
+		}
+		fullLog := p.prefix + log
+		if len(fullLog) > maxWidth {
+			fullLog = fullLog[:maxWidth-4] + cPostfix
+		}
+
+		p.clear()
+		p.output = pushRemove(p.output, fullLog)
+		p.printOutput()
+	}
+	return len(logBytes), nil
+}
+
+func pushRemove(original []string, value string) []string {
+	copy := original[1:] // remove first
+	return append(copy, value)
+}
+
+func (p *previewer) printOutput() {
+	size := len(p.output)
+	count := 0
+
+	tm.ResetLine("")
+	tm.Println(p.displayTitle)
+	// margin after title
+	tm.Println("")
+
+	for count < size {
+		tm.ResetLine("")
+		tm.Println(p.output[count])
+		count++
+	}
+
+	// margin after output
+	tm.Println("")
+	tm.Print(p.footerLine)
+
+	tm.Flush()
```

Comment:
I don't love how this existing package handles the io.Writer that it will write to.  The value is a package global thing, and I think it would be much better if that wasn't the case so we could tie this to the `io.Writer` we inject.

Maybe we just need to bring this logic into this component itself and have us write the ANSI sequences for movement and the like?

I think this will make this a more re-usable and easier to test component.

Reply:
The package has a `Screen` exported field which is used as the screen-buffer and a `Output` exported field.

```
var Screen *bytes.Buffer = new(bytes.Buffer)
var Output *bufio.Writer = bufio.NewWriter(os.Stdout)
```

This 2 fields can be set before calling `Start()` to define the buffer to use (in case we don't want the package to auto-allocate) and the Writer.

Reply:
That provides us some level of control but we've tried to eschew package level state from as many places as we can.

I do wonder if we need to be setting the value of `Output` in the function we register in `registerCommonDependencies`?  Presumably we should set it not to `os.Stdout` but rather `cmd.OutOrStdout()` here.  I suspect in practice it does not matter given how we use the system today, but probably a smart idea anyway.

I think I am okay with this as is (once we do the dependency registration stuff) but it would be good to move to some design where we can more easily ensure that we use the same output writer as the rest of our console abstraction.  I think long term we really are going to want to be able to work in terms of `io.Writers` and `io.Readers` and packages that make it difficult to do that (by making package level state decisions) is going to make that much harder for us.